### PR TITLE
signer/core/apitypes: support more input types for eip-712 encoding

### PR DESCRIPTION
**Pull request by Obtuse7772**
1. Previously, geth would not accept byte arrays (i.e. `[32]byte`) even for arguments where `bytes32` is specified. This makes it terribly inconvenient to work with fixed size byte arrays.
2. Additionally, the Typed Data hasher wouldn't accept anything except strings in place of addresses, which doesn't make much sense. It's an 20 byte array after all. Not to mention the inefficiency in using hexadecimal strings in place of a byte array. I've adjusted it to accept both byte slices and byte arrays.
3. It also would not accept `big.Int`s in certain places, despite it being internally converted to a `big.Int` internally.

This PR solves both problems.

### CODE REVIEW

**Review by ligi — DISMISSED** [review]
Please use reflection and add tests

File: signer/core/apitypes/types.go [hunk 1]
```diff
@@ -419,6 +419,39 @@ func (typedData *TypedData) EncodeData(primaryType string, data map[string]inter
...
+	case [24]byte:
+	case [25]byte:
+	case [26]byte:
+	case [27]byte:
+	case [28]byte:
+	case [29]byte:
+	case [30]byte:
+	case [31]byte:
+	case [32]byte:
+		return v[:], true
```
Comment (holiman) [inline signer/core/apitypes/types.go:454]: Writing it like this does not cause fall-through. The `[31]byte`-case is an empty clause, so it does nothing, then exits the switch, and returns `nil, false` -- that was not what you intended, was it? 
The way to do a 'fall-through' is 
```
case [32]byte, [31]byte, ... 
```
However, this is a bit clunky. A more elegant solution could probably be made  using `reflect`. 
All in all, I think you would have / should have caught this with a testcase. We do like to make changes to these pretty complex internals unless it's properly backed by testcase.

Comment (Obtuse7772) [reply signer/core/apitypes/types.go]: Implemented a different solution using reflect. Added tests for 2, 8 and 16 sized byte arrays, which I think should be enough.

File: signer/core/apitypes/types.go [hunk 2]
```diff
@@ -490,13 +500,34 @@ func parseInteger(encType string, encValue interface{}) (*big.Int, error) {
...
+			return retval, nil
+		}
+		byteSlice, isByteSlice := encValue.([]byte)
+		if isByteSlice {
+			if len(byteSlice) == 20 {
+				retval := make([]byte, 32)
+				copy(retval[12:], byteSlice)
+				return retval, nil
+			} else if len(byteSlice) == 32 {
+				return byteSlice, nil
```
Comment (fjl) [inline signer/core/apitypes/types.go:519]: Why would it be 32 bytes long? Addresses have a size of 20 bytes.

Comment (Obtuse7772) [reply signer/core/apitypes/types.go]: Removed support for this, thought it may be useful.

**Comment by holiman** [comment]
> the Typed Data hasher wouldn't accept anything except strings in place of addresses, which doesn't make much sense. It's an 20 byte array after all. Not to mention the inefficiency in using hexadecimal strings in place of a byte array.

EIP [712 spec](https://eips.ethereum.org/EIPS/eip-712)

> This EIP aims to improve the usability of off-chain message signing for use on-chain. We are seeing growing adoption of off-chain message signing as it saves gas and reduces the number of transactions on the blockchain. Currently signed messages are an opaque hex string displayed to the user with little context about the items that make up the message.

The EIP basically outlines how a human-readable text-based structure is encoded to a canonical form, and hashed/signed. The inputs are typically json inputs, and address would therefore be a hex-encoded string. 

What is the usecase where some different type would be passed into `EncodePrimitiveValue`?

**Comment by holiman** [comment]
Also, please add a testcase for the change to parsing integers. e.g 
```diff
--- a/signer/core/apitypes/signed_data_internal_test.go
+++ b/signer/core/apitypes/signed_data_internal_test.go
@@ -128,6 +128,7 @@ func TestParseInteger(t *testing.T) {
                {"int32", "-123", big.NewInt(-123)},
                {"uint32", "0xff", big.NewInt(0xff)},
                {"int8", "0xffff", nil},
+               {"int32", big.NewInt(-124), big.NewInt(-124)},
        } {
                res, err := parseInteger(tt.t, tt.v)
                if tt.exp == nil && res == nil {
```

**Comment by Obtuse7772** [comment]
> > the Typed Data hasher wouldn't accept anything except strings in place of addresses, which doesn't make much sense. It's an 20 byte array after all. Not to mention the inefficiency in using hexadecimal strings in place of a byte array.
> 
> EIP [712 spec](https://eips.ethereum.org/EIPS/eip-712)
> 
> > This EIP aims to improve the usability of off-chain message signing for use on-chain. We are seeing growing adoption of off-chain message signing as it saves gas and reduces the number of transactions on the blockchain. Currently signed messages are an opaque hex string displayed to the user with little context about the items that make up the message.
> 
> The EIP basically outlines how a human-readable text-based structure is encoded to a canonical form, and hashed/signed. The inputs are typically json inputs, and address would therefore be a hex-encoded string.
> 
> What is the usecase where some different type would be passed into `EncodePrimitiveValue`?

Our use case is making contract transactions that require the same data structure to be used as an input to the typed data hasher and encoded using the ABI at multiple times. Despite the types being the same, it doesn't make sense that some APIs expect different types for the same underlying thing. These data structures are internally generated and not provided from a human.

EIP712 may have been intended for human-readable inputs, but that doesn't imply one shouldn't use programmatically.

File: signer/core/apitypes/types.go [hunk 3]
```diff
@@ -490,13 +500,32 @@ func parseInteger(encType string, encValue interface{}) (*big.Int, error) {
...
+				return retval, nil
+			} else {
+				return nil, errors.New("if a byte slice is passed in, in place of an address, it must be either 20 or 32 bytes long")
+			}
+		}
+		twentyByteArr, isTwentyByteArr := encValue.([20]byte)
+		if isTwentyByteArr {
+			retval := make([]byte, 32)
+			copy(retval[12:], twentyByteArr[:])
+			return retval, nil
```
Comment (holiman) [inline signer/core/apitypes/types.go:526]:
```suggestion
		retval := make([]byte, 32)
		switch val := encValue.(type) {
		case string:
			if common.IsHexAddress(val) {
				copy(retval[12:], common.HexToAddress(val).Bytes())
				return retval, nil
			}
		case []byte:
			if len(val) == 20 {
				copy(retval[12:], val)
				return retval, nil
			}
		case [20]byte:
			copy(retval[12:], val[:])
			return retval, nil
		}
```
IMO a bit clearer, and definitely shorter

Comment (holiman) [reply signer/core/apitypes/types.go:526]: Also, we need tests for these type-conversions, (and preferrably the fail-cases too)

Comment (Obtuse7772) [reply signer/core/apitypes/types.go]: Added tests for this, and adopted your change.

**Review by holiman — APPROVED** [review]
LGTM

**Pull request merged**